README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Build
Status](https://travis-ci.org/bradleyjeck/epanetReader.svg?branch=master)](https://travis-ci.org/bradleyjeck/epanetReader)
[![Coverage
Status](https://codecov.io/gh/bradleyjeck/epanetReader/branch/master/graph/badge.svg)](https://codecov.io/gh/bradleyjeck/epanetReader)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/epanetReader)](https://cran.r-project.org/package=epanetReader)
[![CRAN
version](http://www.r-pkg.org/badges/version/epanetReader)](https://cran.r-project.org/package=epanetReader)

# epanetReader

epanetReader is an R package for reading water network simulation data
in Epanet’s .inp and .rpt formats into R. Some basic summary information
and plots are also provided.

Epanet is a highly popular tool for water network simulation. But, it
can be difficult to access network information for subsequent analysis
and visualization. This is a real strength of R however, and there many
tools already existing in R to support analysis and visualization.

In addition to this README page, information about epanetReader is
available from [Environmental Modelling &
Software](http://www.sciencedirect.com/science/article/pii/S1364815216302870)
[(pdf)](http://bradeck.net/docs/Eck2016epanetReaderEMS.pdf) and [ASCE
Conference
Proceedings](http://ascelibrary.org/doi/abs/10.1061/9780784479865.051)
[(pdf)](http://bradeck.net/docs/Eck2016epanetReader-ewri.pdf).

## Installation

  - the latest released version: install.packages(“epanetReader”)
  - the development version:
    devtools::install\_github(“bradleyjeck/epanetReader”)

## Getting Started

### Network files

Read network information from an .inp file with a similar syntax as the
popular read.table or read.csv functions. Note that the example network
one that ships with Epanet causes a warning. The warning is just a
reminder of how R deals with integer IDs.

``` r
library(epanetReader)
n1 <- read.inp("Net1.inp") 
```

    ## Warning in PATTERNS(allLines): patterns have integer IDs, see ?epanet.inp

    ## Warning in CURVES(allLines): curves have integer IDs, see ?epanet.inp

Retrieve summary information about the network.

``` r
summary(n1)
```

    ## EPANET Example Network 1
    ## A simple example of modeling chlorine decay. Both bulk and
    ## wall reactions are included.
    ## 
    ##             Number
    ## Junctions        9
    ## Tanks            1
    ## Reservoirs       1
    ## Pipes           12
    ## Pumps            1
    ## Quality         11
    ## Coordinates     11
    ## Labels           3

A basic network plot is also available

``` r
plot(n1)
```
![Net 1 plot](https://github.com/bradleyjeck/epanetReader/blob/master/img/Net1inp.png)

The read.inp function returns an object with structure similar to the
.inp file itself. A section in the .inp file corresponds to a named
entry in the list. These entries are accessed using the $ syntax of
    R.

``` r
names(n1)
```

    ##  [1] "Title"       "Junctions"   "Tanks"       "Reservoirs"  "Pipes"      
    ##  [6] "Pumps"       "Valves"      "Demands"     "Patterns"    "Curves"     
    ## [11] "Controls"    "Rules"       "Energy"      "Status"      "Emitters"   
    ## [16] "Quality"     "Sources"     "Reactions"   "Mixing"      "Times"      
    ## [21] "Report"      "Options"     "Coordinates" "Vertices"    "Labels"     
    ## [26] "Backdrop"    "Tags"

Sections of the .inp file are stored as a data.frame or character
vector. For example, the junction table is stored as a data.frame and
retrieved as follows. In this case patterns were not specified in the
junction table and so are marked NA.

``` r
n1$Junctions
```

    ##   ID Elevation Demand Pattern
    ## 1 10       710      0      NA
    ## 2 11       710    150      NA
    ## 3 12       700    150      NA
    ## 4 13       695    100      NA
    ## 5 21       700    150      NA
    ## 6 22       695    200      NA
    ## 7 23       690    150      NA
    ## 8 31       700    100      NA
    ## 9 32       710    100      NA

A summary of the junction table shows that Net1.inp has nine junctions
with elevations ranging from 690 to 710 and demands ranging from 0 to
200. Note that the node ID is stored as a character rather than an
integer or
    factor.

``` r
summary(n1$Junctions)
```

    ##       ID              Elevation         Demand      Pattern       
    ##  Length:9           Min.   :690.0   Min.   :  0.0   Mode:logical  
    ##  Class :character   1st Qu.:695.0   1st Qu.:100.0   NA's:9        
    ##  Mode  :character   Median :700.0   Median :150.0                 
    ##                     Mean   :701.1   Mean   :122.2                 
    ##                     3rd Qu.:710.0   3rd Qu.:150.0                 
    ##                     Max.   :710.0   Max.   :200.0

### Epanet Simulation Results

Results of the network simulation specified in Net.inp may be stored in
Net1.rpt by running Epanet from the command line. Note that the report
section of the .inp file should contain the following lines in order to
generate output readable by this package.

> \[REPORT\]  
> Page 0  
> Links All  
> Nodes All

On windows, calling the epanet executable epanet2d runs the simulation.

    >epanet2d Net1.inp Net1.rpt 
    
    ... EPANET Version 2.0
    
      o Retrieving network data
      o Computing hydraulics 
      o Computing water quality
      o Writing output report to Net1.rpt
    
    ... EPANET completed.

The .rpt file generated by Epanet may be read into R using read.rpt().
The simulation is summarized over junctions, tanks and pipes.

``` r
n1r <- read.rpt("Net1.rpt") 
summary(n1r)
```

    ## Contains node results for  25 time steps 
    ## 
    ## Summary of Junction Results: 
    ##      Demand         Pressure        Chlorine     
    ##  Min.   :  0.0   Min.   :106.8   Min.   :0.1500  
    ##  1st Qu.: 80.0   1st Qu.:116.1   1st Qu.:0.3500  
    ##  Median :120.0   Median :119.8   Median :0.5100  
    ##  Mean   :122.2   Mean   :119.6   Mean   :0.5434  
    ##  3rd Qu.:160.0   3rd Qu.:123.0   3rd Qu.:0.7400  
    ##  Max.   :320.0   Max.   :133.9   Max.   :1.0000  
    ## 
    ## Summary of Tank Results:
    ##      Demand             Pressure        Chlorine    
    ##  Min.   :-1100.000   Min.   :48.22   Min.   :0.590  
    ##  1st Qu.: -660.000   1st Qu.:52.00   1st Qu.:0.660  
    ##  Median :  258.000   Median :55.52   Median :0.750  
    ##  Mean   :   -5.741   Mean   :54.86   Mean   :0.764  
    ##  3rd Qu.:  505.380   3rd Qu.:57.54   3rd Qu.:0.850  
    ##  Max.   : 1029.420   Max.   :60.04   Max.   :1.000  
    ## 
    ## Contains link results for  25 time steps 
    ## 
    ## Summary of Pipe Results:
    ##       Flow             Velocity         Headloss    
    ##  Min.   :-1029.42   Min.   :0.0000   Min.   :0.000  
    ##  1st Qu.:   41.37   1st Qu.:0.3475   1st Qu.:0.110  
    ##  Median :  113.08   Median :0.5700   Median :0.300  
    ##  Mean   :  245.35   Mean   :0.8070   Mean   :0.644  
    ##  3rd Qu.:  237.23   3rd Qu.:1.0075   3rd Qu.:0.755  
    ##  Max.   : 1909.42   Max.   :2.7300   Max.   :3.210  
    ## 
    ## Energy Usage:
    ##   Pump usageFactor avgEfficiency kWh_per_Mgal avg_kW peak_kW dailyCost
    ## 1    9       57.71            75       880.42  96.25   96.71         0

The default plot of simulation results is a map for time period
00:00:00. Note that the object created from the .inp file is a required
argument to make the plot.

``` r
plot( n1r, n1)
```
![Net 1 rpt plot](https://github.com/bradleyjeck/epanetReader/blob/master/img/Net1rpt.png)

In contrast to the treatment of .inp files described above, data from
.rpt files is stored using a slightly different structure than the .rpt
file. The function returns an object (list) with a data.frame for node
results and data.frame for link results. These two data frames contain
results from all the time periods. This storage choice was made to
facilitate time series plots.

Entries in the epanet.rpt object (list) created by read.rpt() are found
using the names() function.

``` r
names(n1r)
```

    ## [1] "nodeResults" "linkResults" "energyUsage"

Results for a chosen time period can be retrieved using the subset
function.

``` r
subset(n1r$nodeResults, Timestamp == "0:00:00")
```

    ##    ID   Demand    Head Pressure Chlorine      note Timestamp timeInSeconds
    ## 1  10     0.00 1004.35   127.54      0.5             0:00:00             0
    ## 2  11   150.00  985.23   119.26      0.5             0:00:00             0
    ## 3  12   150.00  970.07   117.02      0.5             0:00:00             0
    ## 4  13   100.00  968.87   118.67      0.5             0:00:00             0
    ## 5  21   150.00  971.55   117.66      0.5             0:00:00             0
    ## 6  22   200.00  969.08   118.76      0.5             0:00:00             0
    ## 7  23   150.00  968.65   120.74      0.5             0:00:00             0
    ## 8  31   100.00  967.39   115.86      0.5             0:00:00             0
    ## 9  32   100.00  965.69   110.79      0.5             0:00:00             0
    ## 10  9 -1866.18  800.00     0.00      1.0 Reservoir   0:00:00             0
    ## 11  2   766.18  970.00    52.00      1.0      Tank   0:00:00             0
    ##     nodeType
    ## 1   Junction
    ## 2   Junction
    ## 3   Junction
    ## 4   Junction
    ## 5   Junction
    ## 6   Junction
    ## 7   Junction
    ## 8   Junction
    ## 9   Junction
    ## 10 Reservoir
    ## 11      Tank

A comparison with the corresponding entry of the .rpt file, shown below
for reference, shows that four columns have been added to the table.
These pieces of extra info make visualizing the results easier.

``` 
  Node Results at 0:00:00 hrs:
  --------------------------------------------------------
                     Demand      Head  Pressure  Chlorine
  Node                  gpm        ft       psi      mg/L
  --------------------------------------------------------
  10                   0.00   1004.35    127.54      0.50
  11                 150.00    985.23    119.26      0.50
  12                 150.00    970.07    117.02      0.50
  13                 100.00    968.87    118.67      0.50
  21                 150.00    971.55    117.66      0.50
  22                 200.00    969.08    118.76      0.50
  23                 150.00    968.65    120.74      0.50
  31                 100.00    967.39    115.86      0.50
  32                 100.00    965.69    110.79      0.50
  9                -1866.18    800.00      0.00      1.00  Reservoir
  2                  766.18    970.00     52.00      1.00  Tank
```

### Epanet-msx simulation results

Results of a multi-species simulation by Epanet-msx can be read as well.

The read.msxrpt() function creates an s3 object of class epanetmsx.rpt.
Similar to the approach above, there is a data frame for node results
and link results.

## Usage with other packages

### ggplot2

The ggplot2 package makes it easy to create complex graphics by allowing
users to describe the plot in terms of the data. Continuing the Net1
example Here we plot chlorine concentration over time at each node in
the network.

``` r
library(ggplot2)
qplot( data= n1r$nodeResults,  
       x = timeInSeconds/3600, y = Chlorine, 
       facets = ~ID, xlab = "Hour")  
```

![Net 1 Cl plot](https://github.com/bradleyjeck/epanetReader/blob/master/img/Net1cl.png)

### Animation

The animation package is useful for creating a video from successive
plots.

``` r
# example with animation package 
library(animation)

#unique time stamps
ts <- unique((n1r$nodeResults$Timestamp))
imax <- length(ts)

# generate animation of plots at each time step
saveHTML(
  for( i in 1:imax){
    plot(n1r, n1, Timestep = ts[i]) 
  }
)
```

## References

Rossman, L. A. (2000) [Epanet 2 users
manual](http://nepis.epa.gov/Adobe/PDF/P1007WWU.pdf). US EPA,
Cincinnati, Ohio.


NEWS
epanetReader version 0.7.0 (released May 2018)
=============================================
Changes:
* added support for reading energy usage table from rpt files 
* switched from Kmisc::readlines to data.table::fread for faster reading of files
  due to absence of Kmisc on CRAN 


epanetReader version 0.6.2 (released Jan 2018)
=============================================
Changes:
* added is.XXX functions to check class of S3 objects
* update built-in datasets with latest package output and check this with examples
* handle inp file with an empty options section  


epanetReader version 0.4.0 (released Aug 2016)
==============================================
Changes:
* Use readlines from package Kmisc, if available, to speed up reading of large files

* Added write.inp() function for writing inp files from an R object

* Allow page breaks in rpt files to support reading rpt files
  created by Epanet's  Windows GUI
  
* Added citation information for recently published papers

* Added support for reading sections of inp files that were not supported yet

* Handled % sign in rpt file node result header for water trace analysis


epanetReader version 0.3.1 (released Feb 2016)
==============================================

Changes:
* Added read.msxrpt() function to read results of multi-species 
  water quality simulations made with EPANET-MSX

* Added plotSparklineTable() function to generate tables
  of sparklines for exploratory analysis 

* Added data objects Net1 and Net1rpt in rdata format so 
  some examples run without having to parse the text files.

* Changed treatment of some columns of sections of .inp files from
  character to factor to improve summaries (e.g. Pipe Status, Valve Type) 

* Added support to read Status and Demands sections of .inp files 

* Plotting changes for simulation results: fixed bug for plotting with valves, 
  changed aspect ratio to 1 for maps
  




R/epanet.rpt-s3.r
#************************************
#
#  (C) Copyright IBM Corp. 2015
#
#  Author: Bradley J Eck
#
#************************************#


#' Read .rpt file
#' 
#' reads an Epanet .rpt file into R
#' 
#' @export 
#' @param file the name of the file to read
#' @return Returns an epanet.rpt S3 object with two data.frame elements. 
#'
#' \item{nodeResults}{data.frame}
#' \item{linkResults}{data.frame}
#' \item{energyUsage}{data.frame}
#'
#' @details add lines "Page 0", "Links All" and "Nodes All" to the
#'  [REPORT] section of the .inp file to output info to read in
#' with this function
#'
#' In contrast to the treatment of .inp files, data from .rpt
#' files is stored using a slightly different structure than the .rpt file.  The
#' function returns an object (list) with a data.frame for node results and
#' data.frame for link results and a data.frame for energy usage.  The node and
#' link results data frames contain results from all the time periods to
#' facilitate time series plots.
#' 
#' @references Rossman, L. A. (2000). Epanet 2 users manual. US EPA, Cincinnati, Ohio.
#'
#' http://nepis.epa.gov/Adobe/PDF/P1007WWU.pdf
#'
#' @examples
#' # path to Net1.rpt example file included with this package
#' rpt <- file.path( find.package("epanetReader"), "extdata","Net1.rpt") 
#' n1r <- read.rpt(rpt)
#' summary(n1r)
#' names(n1r)
#' 
#' #Results for a chosen time period can be retrieved using the subset function.
#' subset(n1r$nodeResults, Timestamp == "0:00:00")
#' 
#' # time series plot for a nodal value
#' plot( Chlorine ~ timeInSeconds,  
#'       data = subset(n1r$nodeResults, ID == "22"))    
#'
#' # Plotting the epanet.rpt object itself gives a map.
#' # Note that the object created from the .inp file is required.
#' inp <- file.path( find.package("epanetReader"), "extdata","Net1.inp") 
#' n1 <- read.inp(inp)
#' plot( n1r, n1)
#'
#' # Energy Usage table
#' print(n1r$energyUsage)

read.rpt <- function( file ){
 
  return( epanet.rpt(file))	
}


epanet.rpt <- function( file){
  # read all the lines in the file 
  allLines <- read_lines_wrapper(file)
  checkRptFile( allLines ) 
	
  cleanLines <- cleanRptLines(allLines)

  lengthOfCleanLines <- length( cleanLines)
  
  resLines <- grep("Results", cleanLines)
  numTables <- length(resLines)
  
  #create a lits of empty data frames to store each of these
  nodeResList <- list()  
  linkResList <- list()  
  
    # initialize indices for these lists 
  ni <- 0  # node index
  li <- 0  # link index
  
  # go through the tables
  if( numTables > 0 ){  
  for( i in 1:numTables ){
    # get the section  
    sectRange <- .getSectionRange( i, resLines, lengthOfCleanLines)
    sect <- cleanLines[ sectRange$start : sectRange$end ]  
    
    # create a data frame from this section
    df <- .section2df( sect )  

    # decide if it's for link or node results 
    isNODE <- grepl( "Node", sect[1] )
    isLINK <- grepl( "Link", sect[1] )
    
    #add  data to approriate list of data frames 
    if( isNODE ){
      # increment node indexer 
      ni <- ni + 1
      nodeResList[[ ni ]] <- df
    }
    
    if( isLINK ){
      # increment indexer 
      li <- li + 1
      linkResList[[ li ]] <- df
    }
  }}
  
  # combine all these results together 
  nodeResDF <- do.call("rbind", nodeResList )
  linkResDF <- do.call("rbind", linkResList )
 
  
  # add a column specifying the type of node or link
  # use a factor for this 
	
	if( is.null(nodeResDF) == FALSE ) {	
		# make all of them juncs 
		nodeResDF$nodeType <- factor("Junction", 
				levels = c("Junction", "Tank", "Reservoir"))	
		# then identify the tanks and reservoirs 
		tankRows <- which(nodeResDF$note == "Tank")  
		nodeResDF$nodeType[tankRows] <- "Tank"	
		resRows <- which(nodeResDF$note == "Reservoir")
		nodeResDF$nodeType[resRows] <- "Reservoir"
	}
  
  # same idea for the links 
	if( is.null(linkResDF) == FALSE){
		linkTypeLevels <- 	c("Pipe", "Pump", 
				"PSV", "PRV", "PBV", "FCV", "TCV", "GPV")
		
		linkResDF$linkType <- factor("Pipe", levels = linkTypeLevels)
		
		allLinksArePipes <- as.logical(max(is.na(linkResDF$note)))
		
		if( allLinksArePipes == FALSE){
			# take the notes that are not blank and cast to factor 
			rows2cast <-  which( linkResDF$note != "" ) 
			linkResDF$linkType[rows2cast] <- factor(linkResDF$note[rows2cast ], levels = linkTypeLevels)  
		}
	}
  
  # Energy Usage Table 
  egyUsgDF <- getEnergyUsage( cleanLines) 
  
  # and make a list of the to return 
  allResults <- list( nodeResults = nodeResDF, 
                      linkResults = linkResDF,
                      energyUsage = egyUsgDF )
  
  class(allResults) <- "epanet.rpt"

  return( allResults ) 
}



#' Summary of Epanet Simulation Results
#'
#' Provides a basic summary of simulation results 
#'
#' @export
#' @param  object of epanet.rpt class
#' @param ... further arguments passed to summary()
#' @details 
#' Summary of pipe results shows positive and negative
#' values of flow but only positive values of velocity
#' as in the rpt file. 
summary.epanet.rpt <- function( object, ... ){

	###############
	# node results 
	###############
	
	# find the number of time steps reported 
	numNodeTimeSteps <- length(unique(object$nodeResults$Timestamp))
	
	if( numNodeTimeSteps > 0 ){
		nodeCols <- names(object$nodeResults)[c(2,4)]
		n5 <- names(object$nodeResults)[5]
		
		if( is.na(n5) == FALSE ){
			if( n5 != "note"){
				nodeCols <- c(nodeCols, n5)
			}
		}
		
		# silence R CMD check 
		nodeType <- NULL 
		
		juncSmry <- summary(  subset(object$nodeResults, nodeType == "Junction",
						select = nodeCols ) )
		
		tankSmry <- summary(  subset(object$nodeResults, nodeType == "Tank", 
						select = nodeCols ) )
	} else {
		
		juncSmry <- NA
		tankSmry <- NA
	}
	 
	 
	################
	#  link results 
	################
	numLinkTimeSteps <- length(unique(object$linkResults$Timestamp))
    
	if( numLinkTimeSteps > 0 ){
		# silence R CMD check 
		linkType <- NULL 
		
		pipeSmry <- summary(  subset(object$linkResults, linkType == "Pipe", 
						select = c("Flow","Velocity", "Headloss") ) )
	} else {
		
		pipeSmry <- NA
	}
	
	
	# Collect items to return in a list 
	rptSmry <- list( numLinkTimeSteps = numLinkTimeSteps,
			         numNodeTimeSteps = numNodeTimeSteps,
			         juncSummary = juncSmry,
					 tankSummary = tankSmry,
					 pipeSummary = pipeSmry,
					 energyUsage = object$energyUsage ) 
	
	class(rptSmry) <- "summary.epanet.rpt"		 
			
	return( rptSmry)
  
}


#' Print rpt summary
#' 
#' The function prints a summary of simulation results
#' contained in the rpt file. 
#'  
#' @export
#' @param x a summary.epanet.rpt object   
#' @param ... further arguments passed to print 
print.summary.epanet.rpt <- function(x,...){
	# how many timesteps for nodes 
	cat("Contains node results for ", x$numNodeTimeSteps, "time steps \n")
	cat("\n") 

	if( x$numNodeTimeSteps >0 ){
		cat("Summary of Junction Results: \n")
		print(x$juncSummary)
		cat("\n") 
		
		cat("Summary of Tank Results:\n")
		print(x$tankSummary)
		cat("\n") 
	}
	
	
	cat("Contains link results for ", x$numLinkTimeSteps, "time steps \n")
	cat("\n") 

	if( x$numLinkTimeSteps >0 ){
		cat("Summary of Pipe Results:\n")
		print(x$pipeSummary)
		cat("\n") 
	} 
	
	if( !is.null(x$energyUsage)){
	  cat("Energy Usage:\n")
    print(x$energyUsage)
    cat("\n") 
	} 
}


#' Get link quantity at a timestep
#'
#' Extract a table of results for a time step
#' @param rpt epanet.rpt object
#' @param linkQty string with quantity to get
#' @param Timestep string  in h:mm:ss form 
#' @return data.frame with cols "Link" and linkQty
.getLinkQtyAtTime <-function( rpt, linkQty, Timestep){

  if( is.null(linkQty)){
	  return (NULL)
  }	else if( is.na(linkQty)){
	  stop("NA not allowed for linkQty, use NULL instead")
  }	else {
	  # there is some value besides NA or NULL
	  # see if it's present in the results 
	  ok <- match(linkQty, names(rpt$linkResults))
	  if( is.na(ok)){
		  stop("linkQty not present in linkResults")
	  }
	 
	  # validate timestep 
	  ok <- match( Timestep, unique(rpt$linkResults$Timestamp))
	  if( is.na(ok)){
		  msg <- paste("link results not available for", Timestep)
		  stop(msg)
	  } else { 
		  
		  # silence R CMD check
		  Timestamp <- NULL
		  
		  # link results for this timestep		  
		  lqty <- subset( rpt$linkResults,
				  Timestamp == Timestep, 
				  select = c( "ID", linkQty ) )
		  
	  }
	  
      return( lqty)
  }

}



#' Get junc quantity at a timestep
#'
#' Extract a table of results for a time step
#' @param rpt epanet.rpt object
#' @param juncQty string with quantity to get
#' @param Timestep string  in h:mm:ss form 
#' @return data.frame with cols "Node" and juncQty
.getNodeQtyAtTime <- function(rpt,juncQty,Timestep){
# validate juncQty
  if( is.null(juncQty)){ 
	  
	  return (NULL)
  } else if( is.na(juncQty)){
	  stop("NA not allowed for juncQty, use NULL instead")
  }	else {
	  # there is some value besides NA or NULL
	  # see if it's present in the results 
	  ok <- match(juncQty, names(rpt$nodeResults))
	  if( is.na(ok) ){
		  stop("juncQty not present in nodeResults")
	  }
	  
	  # check timestep  
	  ok <- match( Timestep, unique(rpt$nodeResults$Timestamp))
	  if( is.na(ok)){
		  msg <- paste("node results not available for", Timestep)
		  stop(msg)
	  } else { 
		  # silence R CMD check
		  Timestamp <- NULL
		  # extract desired results 
		  juncqty <- subset(rpt$nodeResults, Timestamp == Timestep, select=c("ID", juncQty))
	  }	  
    
	  return( juncqty)
  }
  
}


.plotRptLinks <- function(lqty, inp){
	
	if( is.null(lqty) ){
		# handle the case where we don't want to plot link
		# results but still draw the network structure 
	    plotInpLinks(inp)	
	} else {
		# scale quanity 
		#############
		#  Pipes  
		############# 
		
		if( is.null( inp$Pipes) == FALSE ){
			
			# add coordinates to the pipe table 
			ept <-  expandedLinkTable( inp$Pipes, inp$Coordinates )  
		    
			# add bin info to table 
			# could write merge.expandedLinkTable() but prolly not worth it 
			ept2 <- merge( x = ept, by.x = "ID", y = lqty, by.y = "ID")
	
			# plot the segments 
			graphics::segments( x0 = ept2$x1, y0 = ept2$y1,
		              x1 = ept2$x2, y1 = ept2$y2,
					 lwd = ept2$bin   )  
		}
		
		
		#############
		#  Pumps  
		############# 
		if( is.null( inp$Pumps ) == FALSE ){
			ept <-  expandedLinkTable( inp$Pumps, inp$Coordinates )
			
			# add bin info to table 
			# could write merge.expandedLinkTable() but prolly not worth it 
			ept2 <- merge( x = ept, by.x = "ID", y = lqty, by.y = "ID")
	
			# plot the segments 
			graphics::segments( x0 = ept2$x1, y0 = ept2$y1,
		              x1 = ept2$x2, y1 = ept2$y2,
					 lwd = ept2$bin   )  
			
			graphics::points( ept$midx, ept$midy, pch = 8 ) 
		}
		
		
		#############
		#  Valves 
		############# 
		if( is.null( inp$Valves  )  == FALSE ){
			evt <- expandedLinkTable(inp$Valves, inp$Coordinates)
			# add bin info to table 
			# could write merge.expandedLinkTable() but prolly not worth it 
			ept2 <- merge( x = evt, by.x = "ID", y = lqty, by.y = "ID")
	
			# plot the segments 
			graphics::segments( x0 = ept2$x1, y0 = ept2$y1,
		              x1 = ept2$x2, y1 = ept2$y2,
					 lwd = ept2$bin   )  
			
			graphics::points( evt$midx, evt$midy, pch = 25 ,
					bg="black", col = "black" )  
		}
	}
}

# @param ndqty the node quantity 
# @param inp epanet.inp object 
.plotRptNodes <- function(ndqty, inp){
	
	if( is.null( ndqty)){
		
		plotInpNodes(inp, plot.junctions = TRUE)
		
	} else { 
		
		plotInpNodes(inp, plot.junctions = FALSE) # this gets RES and TANKS
		
		# now add the scaled junctions 
		# add coordinates to junctions 
		jpts <- merge( x = subset(inp$Junctions, select = "ID") , by.x = "ID", all.x = TRUE,
				y = inp$Coordinates, by.y = "Node" )
		
		# add desired results too junc table w coordinates 
		jpts <- merge( x = jpts, by.x = "ID", all.x = TRUE,
				y = ndqty, by.y = "ID")
		
		graphics::points( jpts$X.coord, jpts$Y.coord, pch = 21, bg = 'gray',  
				cex = jpts$bin )  
	}
}

#' plot legend for results
#' 
#' helper function to plot a legend for the simulation results 

.plotRptLegend <- function(juncQty, juncBinfo, linkQty, linkBinfo, legend2.locn){
	# use legend with three values  for each qty  
	if( is.null(juncBinfo)){
		juncTitle = ""
		juncNums <- c("","","")
	    jpch = c(NA,NA,NA)	
		jbg = c(NA,NA,NA)
		jcex = c(NA,NA,NA)
	} else {
		juncTitle = juncQty
		juncNums <- juncBinfo$Labels
		if( length(juncNums) != 3){
			stop(" legend is setup for 3 bins")
		}
		jpch = c(21,21,21) 	
		jbg =  c('gray','gray','gray')
		jcex = c(1,2,3) 
	}

	
	
	if( is.null(linkQty)){
		linkTitle = ""
		linkNums = c("","","")
		llwd = c(NA,NA,NA) 
	} else {
		linkTitle = linkQty
		linkNums <- linkBinfo$Labels
		if( length(linkNums) != 3){
			stop(" legend is setup for 3 bins")
		}
		llwd = c(1,2,3)
	}

	
	# now actually plot the legend 
    graphics::legend( legend2.locn, bty  = 'n', ncol =2 ,
			legend=c(juncNums, linkNums) ,
			title = paste(juncTitle, linkTitle, sep = "   ") ,
			title.adj = 0.5,
			pch = c(jpch,NA,NA,NA),
			pt.bg = c(jbg, NA,NA,NA),
			pt.cex = c(jcex, NA,NA,NA),
			col = 'black',
			lwd = c(NA,NA,NA, llwd)
	)
	
}

#' Plot Simulation Results 
#'
#' Plots simulation results for a single time step in map form  
#'
#' @export
#' @param x  epanet.rpt object 
#' @param inp epanet.inp object associated with x
#' @param Timestep string indicating the time to plot 
#' @param juncQty string specifying which column of x$nodeResults
#'                (Demand, Head, Pressure, Chlorine, etc.) 
#'                to show by circle size at network junctions  
#' @param linkQty string specifying which column of x$linkResults  
#'                (Flow, Velocity, Headloss)
#'                to show by line width on network links  
#' @param legend1.locn string passed to legend() for placing legend of network elements
#' @param legend2.locn string passed to legend() for placing legend of junction and link quantities
#' @param ... further arguments passed to plot 
#' @details juncQty plots and values for Junctions only; Tanks and Reservoirs are not included.
#'          In contrast, linkQty is scaled over all of the link types: Pipes, Pumps & Valves.
#'          These choices aim at a map showing demand at junctions and velocity in links.  
plot.epanet.rpt <- function( x,
		inp,
		Timestep = "0:00:00",
		juncQty = "Demand",
		linkQty = "Velocity",
		legend1.locn = "topright",
		legend2.locn = "topleft",
		... ){
	
	
	# check that coordinates actually exist   
	if( is.null(inp$Coordinates) ){
		stop("network does not have coordinates" )
	}
	
	# node results for this time   
	ndqty <- .getNodeQtyAtTime(x,juncQty,Timestep)
	
	if( is.null(ndqty) == FALSE){
		# silence R CMD check 
		nodeType <- NULL 
		
		allJuncVals <- subset( x$nodeResults, nodeType == "Junction", select = juncQty)
		
		juncBinfo <- binBreaker( allJuncVals[,1], nbin = 3 )
		
		# add bin numbers to ndqty 
		ndqty$bin <- cut( ndqty[,2], breaks = juncBinfo$Breaks, labels = FALSE )
		
	}
	
	#  link results for this time  
	lqty <- .getLinkQtyAtTime(x,linkQty,Timestep)
	
	# Find three bins for link quanity considering
	# simulation results at all times 	
	if( is.null(lqty) == FALSE){
        
		# all the values 
		allLinkVals <- abs(  (subset(x$linkResults, select = linkQty ) ))
		
		# find bin breaks and labels 
		linkBinfo <- binBreaker( allLinkVals[,1], nbin = 3 ) 
	
		# add bin numbers to lqty 
		lqty$bin <- cut( lqty[,2], breaks = linkBinfo$Breaks, labels = FALSE )
		
	}
	
	# create blank plot 
    graphics::par( mar = c(1,1,1,1))
	graphics::plot( range(inp$Coordinates$X.coord),
			range(inp$Coordinates$Y.coord),
			type = 'n', asp = 1,
			xlab = "", xaxt = 'n',
			ylab = "", yaxt = 'n'
	)

  
   .plotRptLinks(lqty, inp)
   
   .plotRptNodes(ndqty,inp)
   
  # legend 1 just to define symbols 
  plotElementsLegend(legend1.locn)
 
  .plotRptLegend(juncQty, juncBinfo, linkQty, linkBinfo, legend2.locn)
}

#' Check if an object has class 'epanet.rpt' 
#'
#' @param x an R object 
#' @export
is.epanet.rpt <- function(x){
  inherits(x,"epanet.rpt")

}
